### PR TITLE
Add GATv2 attention message function and tests

### PR DESCRIPTION
**Pull request by tuanljn**
GATv2 message function, following the discussion in #85.
- MultiHead, one shared MLP per (class, port) for the scores and values
- Softmax over the neighbors, segment-max to avoid float32 overflow
- Unit tests + an integration case with RecurrentCoupler
First of the 5 mechanisms.

### CODE REVIEW

**Review by bdonon — DISMISSED** [review]
It looks excellent! Let's wait for you to test it on Maria's use case, and then we're good 👍

**Review by snouatin-datastorm — DISMISSED** [review]
Looks good!

File: src/energnn/model/coupler/message_passing/message_passing_function.py [hunk 1]
```diff
@@ -164,6 +164,205 @@ def sum_over_ports(__accumulator: jax.Array, mlp_port: tuple[MLP, jax.Array]) ->
...
+        def max_over_edges(accumulator, edge_mlp_tuple):
+            """Accumulates the per-receiver, per-head maximum score over all ports of a hyper-edge set."""
+            hyper_edge_set, mlp_dict = edge_mlp_tuple
+            masked_input, non_fictitious_mask = masked_edge_input(hyper_edge_set)
+
+            def max_over_ports(accumulator, mlp_port):
+                """Accumulates the per-receiver, per-head maximum score over the ports of one hyper-edge set."""
+                mlp, port_array = mlp_port
+                score = jnp.where(non_fictitious_mask, mlp(masked_input)[:, : self.n_heads], neg_inf)
+                return scatter_max(accumulator=accumulator, increment=score, addresses=port_array)
```
Comment (laure-crochepierre) [inline src/energnn/model/coupler/message_passing/message_passing_function.py]: Function mlp(masked_input) is called twice (both in max_over_port / weight_over_ports). Could we do it only once for performance improvement ?

File: src/energnn/model/coupler/message_passing/message_passing_function.py [hunk 2]
```diff
@@ -164,6 +164,205 @@ def sum_over_ports(__accumulator: jax.Array, mlp_port: tuple[MLP, jax.Array]) ->
...
+            def weight_over_ports(accumulator, mlp_port):
+                """Accumulates the softmax-weighted value sum and weight sum over the ports of one hyper-edge set."""
+                value_accumulator, weight_accumulator = accumulator
+                mlp, port_array = mlp_port
+                output = mlp(masked_input)
+                score = output[:, : self.n_heads]
+                value = output[:, self.n_heads :].reshape(-1, self.n_heads, value_size)
+                max_at_receiver = gather(coordinates=max_scores, addresses=port_array)
+                shifted_score = jnp.where(non_fictitious_mask, score - max_at_receiver, jnp.zeros_like(score))
+                weight = jnp.exp(shifted_score) * non_fictitious_mask
```
Comment (laure-crochepierre) [inline src/energnn/model/coupler/message_passing/message_passing_function.py]: weight could also be computed without multiplying by non_fictitious_mask. I suggest considering a compute time test when replacing jnp.zeros_like(score) with - inf

File: src/energnn/model/coupler/message_passing/message_passing_function.py [hunk 3]
```diff
@@ -164,6 +164,205 @@ def sum_over_ports(__accumulator: jax.Array, mlp_port: tuple[MLP, jax.Array]) ->
...
+                            out_size=fused_out_size,
+                            use_bias=self.use_bias,
+                            kernel_init=self.kernel_init,
+                            bias_init=self.bias_init,
+                            final_activation=self.final_activation,
+                            rngs=rngs,
+                        )
+        return nnx.data(mlp_tree)
+
+    def __call__(self, *, graph: Graph, coordinates: jax.Array, get_info: bool = False) -> tuple[jax.Array, dict]:
```
Comment (laure-crochepierre) [inline src/energnn/model/coupler/message_passing/message_passing_function.py:291]: @bdonon Do you have any useful information in mind we could log when get_info=True ?

**Review by laure-crochepierre — COMMENTED** [review]
I have two questions regarding performance, but otherwise it looks good to me! 👍

File: src/energnn/model/coupler/message_passing/message_passing_function.py (hunk 1 above)
Comment (tuanljn) [reply src/energnn/model/coupler/message_passing/message_passing_function.py]: The MLP output is now precomputed: it is evaluated once per (class, port), and both the max pass and the weight pass just read the same score and value. Done in 648eb3d.

File: src/energnn/model/coupler/message_passing/message_passing_function.py (hunk 2 above)
Comment (tuanljn) [reply src/energnn/model/coupler/message_passing/message_passing_function.py]: Done. weight = exp(score - max) with no mask anymore, and the -inf goes in through score.

I moved the mask to value instead. Without it, a receiver that only sees fictitious edges keeps max = -inf, so exp(-inf - (-inf)) = 1 and it stops returning 0 like LocalSum does.

About the compute time test: the forward is unchanged (probably because XLA was already deduplicating the two identical MLP calls), but the backward is faster: 9.0 -> 5.7 ms on 1275 edges (-37%) and 54.4 -> 45.6 ms on 20100 edges (-16%). Outputs are identical.

File: src/energnn/model/coupler/message_passing/message_passing_function.py (hunk 3 above)
Comment (laure-crochepierre) [reply src/energnn/model/coupler/message_passing/message_passing_function.py:291]: As discussed with @tuanljn, this is a broader question than this PR. For now, get_info returns the same output {}.

**Review by bdonon — APPROVED** [review]
Great 👍
